### PR TITLE
BREAKING: Generate System metrics semconv from YAML + move attributes to their own namespace

### LINKED ISSUE #73: Move metric definitions to YAML - System Metrics
Move the metric definitions text from `system-metrics.md` to a respective `system-metrics.yaml` and re-generate the markdown tables.

### PR DESCRIPTION
**Pull request by joaopgrassi**
Closes #73 

This PR does:

- Moves the system metrics definitions to the YAML model file. 
- Because of the "limitations" of tooling AND because the existing metrics share attributes with the same name, this PR also changes the attribute namespaces according with the decision from #51 

## Summary of breaking changes

A summary of the changes[ can be found in this comment](https://github.com/open-telemetry/semantic-conventions/pull/89#issuecomment-1672897047). It can also be seen in other "formats" in the changelog + schema file.

---
~Still need resolution from #51. There are several attributes in system metrics (e.g., state) that are used in different metrics but with different enum values.~

~If we want to keep the markdown output as "short" `state`, we will probably have to see what can be done with the tooling. With the current version of the PR, I have the attributes "namespaced" by using the `attribute_group` feature in order to re-use the attribute across metrics.~

### CODE REVIEW

**Review by jsuereth — APPROVED** [review]
As far as what was done in this PR, this LGTM.

- I agree with the namespaces you chose for attributes, as they line up w/ the metrics.
- I agree with the structure/breakdown in the YAML
- I think the final markdown looks good.


We should do one of two things with this PR:

1. Open a (blocking) bug about providing guidance on how to deal with the breaking changes in system metrics.
2. Advise instrumentation to continue using the older System metrics specification until the system metrics working group has had a chance to walk through their issues and declare this stable.

What does everyone think?

**Comment by arminru** [comment]
This needs to be rebased after https://github.com/open-telemetry/semantic-conventions/pull/99 is merged (or the other way round).

**Comment by joaopgrassi** [comment]
I rebased and added the new metrics introduced in https://github.com/open-telemetry/semantic-conventions/pull/99. 

@jsuereth  As far as the plan goes, I think option 2 makes sense:

> Advise instrumentation to continue using the older System metrics specification until the system metrics working group has had a chance to walk through their issues and declare this stable.

I think once the working group has looked at these metrics and decided they are good, then we can follow up with option 1 and add a issue to track how we will communicate the breaking change and add instructions. 

Additionally, we can also add a Notice to the markdown here, saying these have changed and for instrumentations to continue to use the previous version (point to the last release version). What do you think?

File: docs/system/system-metrics.md
Comment (frzifus) [inline docs/system/system-metrics.md:70]: I assume this should be `system.cpu.logical.count` ?

File: docs/system/system-metrics.md
Comment (frzifus) [inline docs/system/system-metrics.md:96]:
```suggestion
| `system.cpu.logical.count` | int | The logical CPU number [0..n-1] | `1` | Recommended |
```

File: model/metrics/system-metrics.yaml
Comment (frzifus) [inline model/metrics/system-metrics.yaml:46]:
```suggestion
      - ref: system.cpu.logical.count
```

File: model/metrics/system-metrics.yaml
Comment (frzifus) [inline model/metrics/system-metrics.yaml:56]:
```suggestion
      - ref: system.cpu.logical.count
```

**Comment by frzifus** [comment]
Just to verify if I understood it correctly, this PR changes the attribute values of example `system.processes` from e.g.: `running` to `system.processes.running` ?

File: docs/system/system-metrics.md
Comment (arminru) [inline docs/system/system-metrics.md:684]: @frzifus regarding https://github.com/open-telemetry/semantic-conventions/pull/89#issuecomment-1662262781:

It does not change the attribute _values_ ("running", ...) but the attribute _key_ from `status` to `system.processes.status` as per https://github.com/open-telemetry/semantic-conventions/issues/61.

File: model/metrics/system-metrics.yaml
Comment (AlexanderWert) [inline model/metrics/system-metrics.yaml:89]: Similar to the cpu state comment. Since we are listing `cached` here, shouldn't we then also list `buffers` and `shared` here?

And what about `total`? Is it always the case that `total` = `used` + `free` ?

**Review by AlexanderWert — APPROVED** [review]
LGTM. Only two minor questions regarding state value members

File: docs/system/system-metrics.md
Comment (joaopgrassi) [reply docs/system/system-metrics.md:96]: Hum, from the description I thought this was like the "assigned number" of the logical CPU, not how many of them are present. So calling `logical.count` is wrong? I think `system.cpu.logical` is enough maybe?

File: model/metrics/system-metrics.yaml
Comment (joaopgrassi) [reply model/metrics/system-metrics.yaml:30]: @AlexanderWert  Yes, I wasn't sure about it as well, but I think it makes sense. I added 7 of them. Right now we have `interrupt` and I wonder if we should expand to `hardware interrupt` and `software interrupt` as well.

File: model/metrics/system-metrics.yaml
Comment (joaopgrassi) [reply model/metrics/system-metrics.yaml:101]: Good point, I think adding those make sense, they were just missing. Added in [4f88215](https://github.com/open-telemetry/semantic-conventions/pull/89/commits/4f882159f7967f9c5c33bcf08d84e4118a4c3c1f)

File: docs/system/system-metrics.md
Comment (joaopgrassi) [reply docs/system/system-metrics.md]: @frzifus I changed it to `system.cpu.logical` in [19cb403](https://github.com/open-telemetry/semantic-conventions/pull/89/commits/19cb403c98d10562214187631d2b572fa776b358). Let me know if you think this makes sense.

**Comment by joaopgrassi** [comment]
@dmitryax 

> In my opinion, it's hard to track the breaking changes in this PR. I'd suggest separate PR's: 1 for YAML generation with no changes and another one for the adoption of #51.

I definitely agree, but the "problem" is that since multiple metrics use the same attribute key with different semantics there's no way to define them once and share. The only way I see to convert the existing metrics to YAML keeping the attributes as-is is to duplicate them all over in the YAML definition. If folks think that's the best way to go, I can do that. But it's not a trivial amount of work for just being removed right after.

**Comment by joaopgrassi** [comment]
Here's a summary of the changes:

## `system.cpu.*`

### Detailed

|Metrics           |Attributes old               |Attributes new                   |
|------------------|-----------------------------|---------------------------------|
|system.cpu.time   |state<br/>cpu                    |system.cpu.state<br/>system.cpu.logical|
|system.cpu.utilization|state<br/>cpu                    |system.cpu.state<br/>system.cpu.logical|
|system.cpu.physical.count|n/a                          |n/a                              |
|system.cpu.logical.count|n/a                          |n/a                              |

### Summary

|Attribute old |Attribute new |
|---|---|
|`state` |`system.cpu.state`   |
|`cpu`   |`system.cpu.logical` |

## `system.memory.*`

### Detailed

|Metrics           |Attributes old               |Attributes new                   |
|------------------|-----------------------------|---------------------------------|
|system.memory.usage|state                        |system.memory.state              |
|system.memory.utilization|state                        |system.memory.state              |

### Summary

|Attribute old |Attribute new |
|---|---|
|`state` |`system.memory.state`   |

## `system.paging.*`

### Detailed

|Metrics           |Attributes old               |Attributes new                   |
|------------------|-----------------------------|---------------------------------|
|system.paging.usage|state                        |system.paging.state              |
|system.paging.utilization|state                        |system.paging.state              |
|system.paging.faults|type                         |system.paging.type               |
|system.paging.operations|type<br/>direction               |system.paging.type<br/>system.paging.direction|

### Summary

|Attribute old |Attribute new |
|---|---|
|`state` |`system.paging.state`   |
|`type` |`system.paging.type`   |
|`direction` |`system.paging.direction`   |

## `system.disk.*`

### Detailed

|Metrics           |Attributes old               |Attributes new                   |
|------------------|-----------------------------|---------------------------------|
|system.disk.io    |device<br/>direction             |system.device <br/>system.disk.direction|
|system.disk.operations|device<br/>direction             |system.device<br/>system.disk.direction|
|system.disk.io_time|device                       |system.device                    |
|system.disk.operation_time|device<br/>direction             |system.device<br/>system.disk.direction|
|system.disk.merged|device<br/>direction             |system.device<br/>system.disk.direction|

### Summary

|Attribute old |Attribute new |
|---|---|
|`device` |`system.device`   |
|`direction` |`system.disk.direction`   |

## `system.filesystem.*`

### Detailed

|Metrics           |Attributes old               |Attributes new                   |
|------------------|-----------------------------|---------------------------------|
|system.filesystem.usage|device<br/>state<br/>type<br/>mode<br/>mountpoint|system.device<br/>system.filesystem.state<br/>system.filesystem.type<br/>system.filesystem.mode<br/>system.filesystem.mountpoint|
|system.filesystem.utilization|device<br/>state<br/>type<br/>mode<br/>mountpoint|system.device<br/>system.filesystem.state<br/>system.filesystem.type<br/>system.filesystem.mode<br/>system.filesystem.mountpoint|

### Summary

|Attribute old |Attribute new |
|---|---|
|`device`     |`system.device`                |
|`state`      |`system.filesystem.state`      |
|`type`       |`system.filesystem.type`       |
|`mode`       |`system.filesystem.mode`       |
|`mountpoint` |`system.filesystem.mountpoint` |

## `system.network.*`

### Detailed

|Metrics           |Attributes old               |Attributes new                   |
|------------------|-----------------------------|---------------------------------|
|system.network.dropped|device<br/>direction             |system.device<br/>system.network.direction|
|system.network.packets|device<br/>direction             |system.device<br/>system.network.direction|
|system.network.errors|device<br/>direction             |system.device<br/>system.network.direction|
|system.network.io |device<br/>direction             |system.device<br/>system.network.direction|
|system.network.connections|device<br/>protocol<br/>state        |system.device<br/>network.protocol<br/>system.network.state|

### Summary

|Attribute old |Attribute new |
|---|---|
|`device` |`system.device`   |
|`direction` |`system.network.direction`   |
|`protocol`      |`network.protocol`      |
|`state`      |`system.network.state`      |

## `system.processes.*`

### Detailed

|Metrics           |Attributes old               |Attributes new                   |
|------------------|-----------------------------|---------------------------------|
|system.processes.count|status                       |system.processes.status          |
|system.processes.created|n/a                          |n/a                              |

### Summary

|Attribute old |Attribute new |
|---|---|
|`status` |`system.processes.status`   |

File: model/metrics/system-metrics.yaml
Comment (mx-psi) [inline model/metrics/system-metrics.yaml:13]:
```suggestion
  # system.cpu.* metrics and attribute group
```

**Review by mx-psi — COMMENTED** [review]
I second @dmitryax's comments above on https://github.com/open-telemetry/semantic-conventions/pull/89#issuecomment-1671634014, renaming of attributes seems like a separate concern from generating the metrics from YAML. At the very least we should:
1. update the PR title and description to clearly describe both
2. include a link to https://github.com/open-telemetry/semantic-conventions/pull/89#issuecomment-1672897047 on the PR description.
3. add a changelog entry listing the specific changes that #51 implies for this set of metrics

**Comment by joaopgrassi** [comment]
@mx-psi 

1. ✅
2. ✅
3. ✅
4. Add changes to schema file ✅

**Comment by mx-psi** [comment]
Thanks, the changelog entry and summary will be really helpful when addressing this change in the Collector :)

**Comment by joaopgrassi** [comment]
@jsuereth

> We should do one of two things with this PR:
> 
>     1. Open a (blocking) bug about providing guidance on how to deal with the breaking changes in system metrics.
> 
>     2. Advise instrumentation to continue using the older System metrics specification until the system metrics working group has had a chance to walk through their issues and declare this stable.

I created #246 to start the discussion on what/how we want to communicate these breaking changes. Depending on how that goes, we can either merge this PR or leave it open until there's an actual plan.

**Comment by joaopgrassi** [comment]
As part of #212 I also renamed the metrics under `system.processes.*` to `system.process.*`. Updated the changelog/schema file accordingly.

**Comment by dmitryax** [comment]
> As part of https://github.com/open-telemetry/semantic-conventions/issues/212 I also renamed the metrics under system.processes.* to system.process.*. Updated the changelog/schema file accordingly.

Why does it have to go in this PR and cannot be addressed separately? There are much more system metrics with pluralized namespaces like `system.disk.operations`

File: CHANGELOG.md
Comment (dmitryax) [inline CHANGELOG.md:76]: These attributes seem inconsistent. Why `system.network.direction` and `system.network.state` have `system` prefix while `network.protocol` doesn't? Should we have `network.direction` instead of `system.network.direction`? `system.network.state` sounds confusing, I believe it should be `network.connection.status` or `network.connection.state`.

File: CHANGELOG.md
Comment (dmitryax) [inline CHANGELOG.md:78]: Should be `system.process.status`, right?

**Comment by joaopgrassi** [comment]
@dmitryax 

> Why does it have to go in this PR and cannot be addressed separately?

I added to this PR to reduce the amount of breaking changes PR, since this is already breaking. I think keeping them in a single PR makes it easier to discover/manage the breaking changes, but if the majority prefers, I can split, no problem. 

> There are much more system metrics with pluralized namespaces like `system.disk.operations`

The other metrics here don't need to be changed, because they fall under the current recommendations in [Pluralization](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/README.md#pluralization)

> Metric names SHOULD NOT be pluralized, unless the value being recorded represents discrete instances of a [countable quantity](https://en.wikipedia.org/wiki/Count_noun). Generally, the name SHOULD be pluralized only if the unit of the metric in question is a non-unit (like {fault} or {operation}).

> Examples:

> - `system.filesystem.utilization`, `http.server.duration`, and `system.cpu.time` should not be pluralized, even if many data points are recorded.
> - `system.paging.faults`, `system.disk.operations`, and `system.network.packets` should be pluralized, even if only a single data point is recorded.

I changed `system.processes.` because that was a namespace and not just a metric. The rest that are still pluralized all have custom unit names that aligns with the current recommendations above:

```
system.paging.operations -> Unit {operation}
system.disk.operations -> Unit {operation}
system.network.packets -> Unit {packet}
system.network.errors -> Unit {error}
system.network.connections -> {connection}
```

File: CHANGELOG.md
Comment (joaopgrassi) [reply CHANGELOG.md]: Yes, thanks for spotting it. It was also wrong in other places. Fixed it everywhere.

**Comment by dmitryax** [comment]
> I added to this PR to reduce the amount of breaking changes PR, since this is already breaking. I think keeping them in a single PR makes it easier to discover/manage the breaking changes, but if the majority prefers, I can split, no problem.

I personally see this approach as harder to follow. Individual PRs per a dedicated change are easier to follow IMO. Also, I think the YAML PR will be merged much faster as the first step. If we keep adding changes to this PR, it's harder to review, a higher chance to miss something important, and likely much longer to merge.

> The other metrics here don't need to be changed, because they fall under the current recommendations in [Pluralization](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/README.md#pluralization)

That makes sense, thanks for clarifying, I missed that. I still have some comments about that particular change, https://github.com/open-telemetry/semantic-conventions/issues/212#issuecomment-1685722874 to discuss that separately

**Comment by joaopgrassi** [comment]
> I personally see this approach as harder to follow. Individual PRs per a dedicated change are easier to follow IMO. Also, I think the YAML PR will be merged much faster as the first step. If we keep adding changes to this PR, it's harder to review, a higher chance of missing something important, and likely much longer to merge.

@dmitryax that's fair and I generally agree. So, you would prefer to postpone the "pluralization" of the metric names to a different PR, and keep this as simply modifying the attribute names + moving to YAML?

**Comment by dmitryax** [comment]
I would still prefer "moving to YAML" as a separate PR that can be merged pretty quickly I believe. Maybe you can open another PR specifically for moving YAML while keeping this one open for reviewing attribute renaming? Once the YAML PR is merged this one can be rebased on top of it. Removing pluralization from namespaces should be also a separate PR I think

**Comment by joaopgrassi** [comment]
@dmitryax I will see what I can do about it. The issue is because of the reasons I mentioned above (tooling limitations/constraints) it's not easy to replicate the current state in YAML without doing some extra work/lots of duplication that will be thrown away right after.

**Comment by joaopgrassi** [comment]
@dmitryax OK I tried splitting and simply move the existing metrics/attributes **as is** to YAML and it's not really possible. Here is the breakdown of things I tried and didn't work:

- Define each attribute under a metric. Such as `state` under each metric instead of re-using via `attribute-group`. The tool complains with: 

```
Attribute state of Semantic convention 'metric.system.memory.usage'is already defined in {group_by_fqn.get(attr.fqn)}.
Attribute state of Semantic convention 'metric.system.memory.utilization'is already defined in {group_by_fqn.get(attr.fqn)}.
Attribute state of Semantic convention 'metric.system.paging.usage'is already defined in {group_by_fqn.get(attr.fqn)}.
Attribute state of Semantic convention 'metric.system.paging.utilization'is already defined in {group_by_fqn.get(attr.fqn)}.
```

- I tried then to define `state` in a metric and use `ref` in others, but the same problem happens. Turns out there's already a `state` attribute registered at the [top-level namespace for database](https://github.com/open-telemetry/semantic-conventions/blob/main/model/metrics/database-metrics.yaml#L6) (which we should also change..). I can override the `brief/note` but **I can't** override the members. This will also happen for other attributes that are enums in the existing system metrics conventions.

Another approach I could take is: Modify the current markdown and introduce the breaking changes there first. THEN move them to YAML. 

I honestly don't think this option makes anything better, it will be two big PRs again. Given I already have this PR ready and all works I think we should go forward with it.

I just want to make it clear that I understand your concerns, but it seems with the current tools we can't do much about it. I can at least revert the "pluralization" changes and this PR is only about the attributes then. Not sure if that makes anything better.. I think I prefer to have all breaking changes in a single PR.

**Comment by jamesmoessis** [comment]
Having worked with the semconv generator, I sympathise with the issues @joaopgrassi is running into. I agree it is inconvenient but necessary that the attributes should be renamed in this PR. Trying to get around this would take more effort than it's worth.

**Review by jamesmoessis — APPROVED** [review]

File: CHANGELOG.md
Comment (mx-psi) [reply CHANGELOG.md:74]: IMO consistency trumps the better naming here so I would keep this as `system.disk.direction`

**Comment by joaopgrassi** [comment]
Discussed today in the System semconv SIG meeting.

- We want to merge this as is, to unblock things
- @dmitryax has some concerns over some of the attributes. Will open follow up issues so we can address those individually
- About the migration plan: The SIG agrees with the plan in https://github.com/open-telemetry/semantic-conventions/issues/246#issuecomment-1699446543. Mostly will follow the transition plan done for HTTP.

**Review by mx-psi — APPROVED** [review]

File: CHANGELOG.md
Comment (joaopgrassi) [reply CHANGELOG.md:75]: Alright, let's keep it as is then 👍

File: docs/system/system-metrics.md
Comment (joaopgrassi) [reply docs/system/system-metrics.md]: @frzifus hi! Did you had a chance to look at my reply here? Does that resolve your concern?

File: model/metrics/system-metrics.yaml
Comment (joaopgrassi) [reply model/metrics/system-metrics.yaml]: Let's combine the discussion in a single thread https://github.com/open-telemetry/semantic-conventions/pull/89#discussion_r1282760528

File: model/metrics/system-metrics.yaml
Comment (joaopgrassi) [reply model/metrics/system-metrics.yaml]: Let's combine the discussion in a single thread https://github.com/open-telemetry/semantic-conventions/pull/89#discussion_r1282760528

**Comment by dmitryax** [comment]
@joaopgrassi I submitted https://github.com/open-telemetry/semantic-conventions/issues/308. We should probably do the same for other metrics

**Review by dmitryax — APPROVED** [review]
LGTM to unblock further changes

File: CHANGELOG.md
Comment (joaopgrassi) [reply CHANGELOG.md:88]: Marking as "resolved" as we will deal with it in a follow-up #308

File: docs/system/system-metrics.md
Comment (frzifus) [reply docs/system/system-metrics.md]: @joaopgrassi Is it in that case a duplicate of https://github.com/open-telemetry/semantic-conventions/pull/89/files#diff-a5f4e0108cc187e4aeaded0612532f0f3d65c0261854c66389273b57c1faa0cbR64-R78?

Comment (mx-psi) [reply docs/system/system-metrics.md]: > @joaopgrassi Is it in that case a duplicate of https://github.com/open-telemetry/semantic-conventions/pull/89/files#diff-a5f4e0108cc187e4aeaded0612532f0f3d65c0261854c66389273b57c1faa0cbR64-R78?

AIUI these are different: `system.cpu.logical` is a metric _attribute_ name while `system.cpu.logical.count` is a metric name

Comment (joaopgrassi) [reply docs/system/system-metrics.md]: Yeah they are different, but I can maybe see it being confusing. Maybe we can change the attribute to `system.cpu.logical_number`?

Comment (frzifus) [reply docs/system/system-metrics.md]: ok, I see - thanks for clarification! I like `system.cpu.logical_number` what do you think?

Comment (joaopgrassi) [reply docs/system/system-metrics.md]: Alright, I will change it then!

**Review by frzifus — APPROVED** [review]

File: docs/system/system-metrics.md
Comment (frzifus) [reply docs/system/system-metrics.md]: Awesome

Comment (joaopgrassi) [reply docs/system/system-metrics.md]: Done in [e7bbf62](https://github.com/open-telemetry/semantic-conventions/pull/89/commits/e7bbf62a44cbac66de1d202ce3a9a1962da792db)

**Review by mx-psi — APPROVED** [review]
cc @open-telemetry/semconv-system-approvers the `cpu` attribute got renamed to `system.cpu.logical_number` (see more in https://github.com/open-telemetry/semantic-conventions/pull/89#discussion_r1281938102 ), please speak out if you disagree with the change

**Review by ChrsMark — APPROVED** [review]

**Comment by ChrsMark** [comment]
> cc @open-telemetry/semconv-system-approvers the `cpu` attribute got renamed to `system.cpu.logical_number` (see more in [#89 (comment)](https://github.com/open-telemetry/semantic-conventions/pull/89#discussion_r1281938102) ), please speak out if you disagree with the change

Don't want to block the merge of this but shouldn't the `system.cpu.logical_number` be under the `host.` namespace since it's a resource attribute? So it would be `host.cpu.logical_number` or `host.cpu.id` similar to what we have for the `host.id`? 

That's related to https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/26533#discussion_r1322766231.

In order to not keep this one blocked we can revisit this as part of System SIG along with the others' attributes stabilization.

**Comment by joaopgrassi** [comment]
> In order to not keep this one blocked we can revisit this as part of System SIG along with the others' attributes stabilization.

@ChrsMark I added your concerns to #308 where we are grouping things to visit. Thanks!

**Comment by joaopgrassi** [comment]
Discussed how to move forward with this PR in the semconv sig meeting (Sept 11th). 

- The system semconv WG agrees with this PR and gave the approvals to merge. 
- Migration plan: I will send a PR to update the system-metrics.md document and add a section similar to what we have in the HTTP about instrumentations to keep using the old version while it's being stabilized. System semconv SIG will then fill in the gaps of the actual migration plan

@open-telemetry/specs-semconv-approvers @open-telemetry/specs-semconv-maintainers this good to merge!

**Pull request merged**
